### PR TITLE
[Bugfix] Implement upgrade-aware controller ordering for FE/BEs/CNs

### PR DESCRIPTION
**Pull request by jmjm15x**
# Description

Add upgrade sequencing control to the StarRocks Kubernetes Operator to ensure proper component ordering during both initial deployments and upgrades. Previously, the operator always used FE-first ordering, which is correct for initial deployments but incorrect for upgrades. According to StarRocks guidelines:

- **Initial Deployment:** FE → BEs/CNs (FE must be leader before workers join)
- **Cluster Upgrades:** BEs/CNs → FE (data nodes upgraded before metadata nodes)

From [official documentation](https://docs.starrocks.io/docs/deployment/upgrade/#upgrade-procedure):

> **Upgrade procedure**
> By design, BEs and CNs are backward compatible with the FEs. Therefore, you need to upgrade BEs and CNs first and then FEs to allow your cluster to run properly while being upgraded. Upgrading them in an inverted order may lead to incompatibility between FEs and BEs/CNs, and thereby cause the service to crash.

## Solution

Implemented a comprehensive upgrade detection and sequencing mechanism with robust component readiness validation to prevent premature progression between components.

#### 1. **Upgrade Detection (`isUpgrade()`)**
Detects upgrade scenarios by checking if StatefulSets exist with pending changes, compares `Generation` vs `ObservedGeneration` to detect spec changes.

**Why this approach?**
- Simple and reliable, uses Kubernetes native generation tracking
- Works for any spec change (images, resources, configs)
- Handles transient states, correctly identifies upgrade during rollout progress

#### 2. **Controller Ordering (`getControllersInOrder()`)**
Dynamically switches controller execution order based on deployment scenario:
- **Upgrade scenario:** `[be, cn, fe, feproxy]` (BE-first ordering)
- **Initial deployment:** `[fe, be, cn, feproxy]` (FE-first ordering)

#### 3. **Component Readiness Validation (`isComponentReady()`)**
Multi-layer validation to prevent premature component progression by avoiding race conditions, ensuring rollout stability, and providing enhanced logging for debugging.

### Logic Flow

Implements waiting logic directly in the reconciliation loop:

```
Reconcile() called
    ↓
Get controllers in order based on isUpgrade()
    ↓
For each controller in order:
    ↓
    ├─ If upgrade && feController
    │   └─ Check BE/CN ready? → If NO, wait and requeue
    │
    ├─ Sync controller (create/update resources)
    │
    ├─ If initial && feController  
    │   └─ Check FE ready? → If NO, wait and requeue
    │
    └─ If upgrade && (beController || cnController)
        └─ Check component ready? → If NO, wait and requeue
```

#### Test Case 1: Initial FE+BE Deployment (v3.1.0)
**Expected:** FE-first ordering (FE must be ready before BE starts)  

```
Timeline:
  T0:       FE Pod Created - 2025-10-09 07:29:58 UTC
  T0+48s:   BE Pod Created - 2025-10-09 07:30:46 UTC

Operator Logs:
  - "initial deployment: waiting for FE to be ready before creating BE/CN"
  - Component progression: feController → wait for FE ready → beController

Verification:
  - FE StatefulSet created first
  - BE StatefulSet created 48 seconds later (after FE became ready)
```

#### Test Case 2: Version Upgrade (v3.1.0 → v3.1.8)
**Expected:** BE-first ordering (BE must complete before FE starts)  

```
Timeline:
  T0:       BE Pod Created - 2025-10-09 07:44:45 UTC
  T0+6s:    FE Pod Created - 2025-10-09 07:44:51 UTC

Operator Logs:
  - "component not ready: StatefulSet spec change not yet observed"
  - "component not ready: StatefulSet rollout in progress"
  - "component not ready: no ready endpoints"
  - "upgrade: waiting for component rollout to complete before proceeding"

Verification:
  - BE StatefulSet updated first (detected generation change)
  - BE pod rolled out to v3.1.8
  - FE update waited for BE rollout completion
  - FE pod rolled out to v3.1.8 after BE was ready
```

#### Test Case 3: Configuration Change (Memory: 2Gi → 4Gi)
**Expected:** BE-first ordering (config changes treated as upgrades)  

```
Timeline:
  T0:       BE Pod Created - 2025-10-09 07:47:05 UTC
  T0+6s:    FE Pod Created - 2025-10-09 07:47:11 UTC

Operator Logs:
  - "component not ready: no ready endpoints"
  - "upgrade: waiting for component rollout to complete before proceeding"

Verification:
  - Configuration change correctly detected as upgrade scenario
  - BE rolled out first with new memory limits (4Gi)
  - FE waited for BE readiness before rolling out
  - Both components running with 4Gi memory
```

## Verification

```bash
# Check initial deployment uses FE-first (wait message)
kubectl logs -n <namespace> deployment/kube-starrocks-operator | grep "initial deployment: waiting for FE"

# Check upgrade uses BE-first (wait message)
kubectl logs -n <namespace> deployment/kube-starrocks-operator | grep "upgrade: waiting for component rollout"

# Verify component readiness details
kubectl logs -n <namespace> deployment/kube-starrocks-operator | grep "component not ready"

# Check StatefulSet rollout sequence with timestamps
kubectl get statefulsets -n <namespace> -o json | jq -r '.items[] | "\(.metadata.name): \(.status.currentRevision) -> \(.status.updateRevision)"'

# Verify pod creation/recreation timestamps
kubectl get pods -n <namespace> -o custom-columns=NAME:.metadata.name,CREATED:.metadata.creationTimestamp
```

# Checklist

For operator, please complete the following checklist:

- [x] run `make generate` to generate the code.
- [x] run `golangci-lint run` to check the code style (0 issues).
- [x] run `make test` to run UT (all controller tests passing).
- [x] run `make manifests` to update the yaml files of CRD.

For helm chart, please complete the following checklist:

- [ ] make sure you have updated the [values.yaml](../helm-charts/charts/kube-starrocks/charts/starrocks/values.yaml)
  file of starrocks chart.
- [ ] In `scripts` directory, run `bash create-parent-chart-values.sh` to update the values.yaml file of the parent
  chart( kube-starrocks chart).

### CODE REVIEW

File: pkg/controllers/controllers.go [hunk 1]
```diff
@@ -38,6 +47,51 @@ func SetupClusterReconciler(mgr ctrl.Manager) error {
...
+) []subcontrollers.ClusterSubController {
+	if isUpgrade(ctx, client, cluster) {
+		return []subcontrollers.ClusterSubController{be, cn, fe, feproxy}
+	}
+
+	return []subcontrollers.ClusterSubController{fe, be, cn, feproxy}
+}
+
+// isUpgrade determines if the current reconciliation is an upgrade scenario.
+func isUpgrade(ctx context.Context, kubeClient client.Client, cluster *srapi.StarRocksCluster) bool {
```
[the remark this answers is not in the document]
Comment (jmjm15x) [reply pkg/controllers/controllers.go]: I’m leaning toward treating any image change as an upgrade in this PR (will push the commit soon)

Downgrade is rare and has strict constraints, so adding version comparison logic now would introduce edge cases and complicate testing. I’d prefer to handle downgrade in a separate PR with proper e2e tests.

Docs suggest downgrade involves major breaking changes, making backward compatibility tricky. Even with semver, version checks aren’t foolproof—human error or provider differences can break assumptions.

[the remark this answers is not in the document]
Comment (jmjm15x) [reply pkg/controllers/controllers.go]: @yandongxiao pushed a new commit to limit upgrade to image changes.

**Comment by yandongxiao** [comment]
Another question: I think this PR should exclude the third PR.

File: pkg/controllers/starrockscluster_controller.go [hunk 2]
```diff
@@ -86,10 +91,32 @@ func (r *StarRocksClusterReconciler) Reconcile(ctx context.Context, req ctrl.Req
...
+
 	// subControllers reconcile for create or update component.
-	for _, rc := range r.Scs {
+	for _, rc := range controllers {
 		kvs := []interface{}{"subController", rc.GetControllerName()}
-		logger.Info("sub controller sync spec", kvs...)
+		controllerName := rc.GetControllerName()
+
+		// During upgrades, check BE and CN are ready BEFORE syncing FE
+		if isUpgradeScenario && controllerName == "feController" {
```
Comment (yandongxiao) [inline pkg/controllers/starrockscluster_controller.go]: `r.FeController.GetControllerName()` is better.

File: pkg/controllers/starrockscluster_controller.go [hunk 3]
```diff
@@ -86,10 +91,32 @@ func (r *StarRocksClusterReconciler) Reconcile(ctx context.Context, req ctrl.Req
...
+			// Check BE readiness if BE exists in spec
+			if src.Spec.StarRocksBeSpec != nil && !isComponentReady(ctx, r.Client, src, "be") {
+				logger.Info("upgrade: waiting for BE rollout to complete before updating FE",
+					"controller", controllerName)
+				return ctrl.Result{}, nil
+			}
+			// Check CN readiness if CN exists in spec
+			if src.Spec.StarRocksCnSpec != nil && !isComponentReady(ctx, r.Client, src, "cn") {
```
Comment (yandongxiao) [inline pkg/controllers/starrockscluster_controller.go]: `componentTypeCN` should be better

File: pkg/controllers/starrockscluster_controller.go [hunk 4]
```diff
@@ -98,9 +125,18 @@ func (r *StarRocksClusterReconciler) Reconcile(ctx context.Context, req ctrl.Req
 			}
 			return requeueIfError(err)
 		}
+
+		// After syncing, check if we need to wait for this component to be ready before proceeding
+		// Initial deployment: Wait for FE to be ready before creating BE/CN
+		if !isUpgradeScenario && controllerName == "feController" {
```
Comment (yandongxiao) [inline pkg/controllers/starrockscluster_controller.go]: `r.FeController.GetControllerName()` should be better.

Comment (jmjm15x) [reply pkg/controllers/starrockscluster_controller.go]: done

**Comment by jmjm15x** [comment]
> Another question: I think this PR should exclude the third PR.

@yandongxiao what do you mean by 3rd PR?

**Comment by yandongxiao** [comment]
> @yandongxiao I noticed a typo in my CLA signature, that's why it's pending. Could you revoke it so I can sign again?

@kevincai

**Comment by yandongxiao** [comment]
> > Another question: I think this PR should exclude the third PR.
> 
> @yandongxiao what do you mean by 3rd PR?

Sorry, there are errors in what I wrote. Now the PR contains four commits, but one commit is from yandongxiao.

**Comment by jmjm15x** [comment]
> > > Another question: I think this PR should exclude the third PR.
> > 
> > 
> > @yandongxiao what do you mean by 3rd PR?
> 
> Sorry, there are errors in what I wrote. Now the PR contains four commits, but one commit is from yandongxiao.

I rebased from main and I think that's the reason for the 3rd commit

File: pkg/controllers/controllers.go [hunk 5]
```diff
@@ -38,6 +47,109 @@ func SetupClusterReconciler(mgr ctrl.Manager) error {
...
+	feSts := &appsv1.StatefulSet{}
+	feExists := kubeClient.Get(ctx, types.NamespacedName{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Name + "-fe",
+	}, feSts) == nil
+
+	beSts := &appsv1.StatefulSet{}
+	beExists := kubeClient.Get(ctx, types.NamespacedName{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Name + "-be",
```
Comment (yandongxiao) [inline pkg/controllers/controllers.go]: use `load.Name(cluster.Name, cluster.Spec.StarRocksBeSpec)`

File: pkg/controllers/controllers.go [hunk 6]
```diff
@@ -82,6 +194,109 @@ func SetupWarehouseReconciler(mgr ctrl.Manager, namespace string) error {
...
+			return true
+		}
+		serviceName = rutils.ExternalServiceName(cluster.Name, cluster.Spec.StarRocksBeSpec)
+		statefulSetName = cluster.Name + "-be"
+	case componentTypeCN:
+		if cluster.Spec.StarRocksCnSpec == nil {
+			return true
+		}
+		serviceName = rutils.ExternalServiceName(cluster.Name, cluster.Spec.StarRocksCnSpec)
+		statefulSetName = cluster.Name + "-cn"
```
Comment (yandongxiao) [inline pkg/controllers/controllers.go]: the same issuse

File: pkg/controllers/starrockscluster_controller.go [hunk 7]
```diff
@@ -98,9 +125,18 @@ func (r *StarRocksClusterReconciler) Reconcile(ctx context.Context, req ctrl.Req
 			}
 			return requeueIfError(err)
 		}
+
+		// After syncing, check if we need to wait for this component to be ready before proceeding
+		// Initial deployment: Wait for FE to be ready before creating BE/CN
+		if !isUpgradeScenario && controllerName == r.FeController.GetControllerName() {
```
Comment (yandongxiao) [inline pkg/controllers/starrockscluster_controller.go]: This brings up a potential issue, e.g. if FE fails to start due to certain reasons, such as excessive metadata or a long startup time, causing the probe to fail. In this case, users would surely want to modify the probe time, but the logic here prevents it.

Comment (yandongxiao) [reply pkg/controllers/starrockscluster_controller.go]: In the sub controller logic, there is a `fe.CheckFEReady(ctx, be.Client, src.Namespace, src.Name)` check, if FE is not ready, BE/CN will stop doing reconcile.

**Review by yandongxiao — CHANGES_REQUESTED** [review]
